Show popup when node name extends beyond drawer width

brandon1024 opened the pull request:
In cases where the node name extends beyond the width of the drawer
window (often the case with long file names or a deeply nested tree),
show a popup to help the user view the full node name.

Closes #395 

![Screencast from 2022-03-04 08_52_26 AM (1)](https://user-images.githubusercontent.com/22732449/156767251-b4bb426b-fa6f-40ce-a042-7d9b74277c06.gif)

## Discussion
These changes have been well tested in my environment (`VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Oct 01 2021 01:51:08)`), but I haven't tested these changes against newer versions of vim. I also haven't tested these changes in Neovim. I gather that popup functionality in Neovim is in flux at the moment. I added a guard to check if `popup_create` exists, which may be implemented by a plugin like `plenary.nvim`, but I'm not sure if that will work properly in neovim or not. If someone could help me test that in Neovim, I would greatly appreciate!

It was a bit tricky making sure that the popup lines up exactly where I want it, especially when nodes/leafs are prefixed with multi-byte unicode characters. You'll see the blob of cursor calculations in this PR, haha.

I'm new to this project, so feedback is appreciated!

lambdalisue commented on autoload/fern/internal/drawer/hover_popup.vim:
```diff
@@ -0,0 +1,46 @@
+  let l:col_offset = wincol() - getcursorcharpos()[2] + l:labelbegin
+
+  call popup_create(l:node.label, #{
+    \ line: 'cursor',
```
Please follow the default value of `g:vim_indent_cont` (6 spaces)

lambdalisue commented on doc/fern.txt:
```diff
@@ -411,7 +411,12 @@ VARIABLE						*fern-variable*
+
+*g:fern#disable_drawer_hover_popup*
+	Set 1 to disable popups shown when the name of a node extends beyond
+	the width of the drawer.
```
Neovim is not supported by this PR so please note that.

lambdalisue commented on autoload/fern/internal/drawer/hover_popup.vim:
```diff
@@ -0,0 +1,46 @@
+    return
+  endif
+
+  if !exists('*popup_create')
```
Please warn users in this case for

1. Neovim is not supported (because popup_create does not exist)
2. This warning message is suppressed by `g:fern#disable_drawer_hover_popup`

brandon1024 replied:
What's the best way to notify the user? I assumed just a warn-level message with `fern#logger#warn()` would be sufficient, but if there's a better way I'll change!

lambdalisue replied:
`fern#logger#warn()` is OK

lambdalisue reviewed (changes requested):
Thanks for the contribution :tada:

brandon1024 commented on autoload/fern/internal/drawer/hover_popup.vim:
```diff
@@ -0,0 +1,46 @@
+endfunction
+
+function! s:cursor_moved_event() abort
+  let l:helper = fern#helper#new()
```
Can do!

brandon1024 commented on autoload/fern/internal/drawer/hover_popup.vim:
```diff
@@ -0,0 +1,46 @@
+  let l:labelbegin = l:labelbegin < 0 ? 0 : l:labelbegin
+  let l:col_offset = wincol() - getcursorcharpos()[2] + l:labelbegin
+
+  call popup_create(l:node.label, #{
```
It was introduced in [4c6d90458](https://github.com/vim/vim/commit/4c6d90458baae843463f930fdc3fe4a7a2191d27) (patch 8.1.1705: using ~{} for a literal dict is not nice, 2019-07-16), but it was only introduced in v0.6.1 of Neovim. I'm going to revert to using the normal dictionary syntax though, in case anyone happens to have a functioning `popup_create()` in Neovim.

brandon1024 commented on autoload/fern/internal/drawer/hover_popup.vim:
```diff
@@ -0,0 +1,46 @@
+  let l:line = getline('.')
+  let l:labelbegin = charidx(l:line, strridx(l:line, l:node.label))
+  let l:labelbegin = l:labelbegin < 0 ? 0 : l:labelbegin
+  let l:col_offset = wincol() - getcursorcharpos()[2] + l:labelbegin
```
I'm kinda struggling on writing these tests. This might take me a bit.

lambdalisue replied:
OK, then push your changes without tests.

brandon1024 replied:
I think I'm starting to get some traction now :-) Thanks for being patient!

brandon1024 replied:
Alright, I've got some tests written but they are currently failing and I'm not sure where to go from here. Hoping you might be able to give some guidance.

I'm not too familiar with `vim-themis`, but I gather that the tests are executed in a "headless" sort of environment. `calculate_node_char_offset()` uses `screenpos()` to calculate the offsets correctly, but because the tests don't actually run with a full-featured tty, it returns all zeroes and doesn't produce the correct values in test.

I could hack around it by accepting the result from `screenpos()` as an argument but that feels icky. How would you approach this problem?

brandon1024 commented:
I remembered that the drawer can be opened on the right side too, and this feature doesn't respond as I would expect it to in this case. I'm making some minor tweaks and should have something shortly.

lambdalisue commented:
Please re-request review when ready.

brandon1024 commented:
Woops, my bad. Addressed feedback but need some help/guidance on the tests (see comment [here](https://github.com/lambdalisue/fern.vim/pull/398#discussion_r821228925)).

lambdalisue commented on autoload/fern/internal/drawer/hover_popup.vim:
```diff
@@ -0,0 +1,63 @@
+endfunction
+
+function! s:show_popup(helper)
+  let node = a:helper.sync.get_cursor_node()
```
You need to follow vint.

brandon1024 replied:
Woops, should be good now :+1:

lambdalisue commented on test/fern/internal/drawer/hover_popup.vimspec:
```diff
@@ -0,0 +1,80 @@
+    End
+  End
+End
+
```
OK. It's a bit tough to review this test so please remove it.

lambdalisue commented:
Thanks for you contribution 🎉

lambdalisue commented:
Now Neovim is supported as well (https://github.com/lambdalisue/fern.vim/pull/403)
